### ISSUE TITLE
send/receive fails on "archive"

### ISSUE THREAD

**Issue by ariep** [digint/btrbk#107]
Hi, I'm trying to use `btrbk archive` to copy snapshots created by btrbk to another computer, as a secondary backup. This is with btrbk version 0.23.3.

This starts out OK, copying three snapshots. But it fails on the fourth one, see output below ("selfukpi" is the primary backup computer, "fukpi" is the secondary one).

```
root@selfukpi:~# btrbk archive -c /etc/btrbk/btrbk.ansible.conf /mnt/data/backup/pook/home/ fukpi:/mnt/data/backups/pook/home/
WARNING: [send/receive] (send=/mnt/data/backup/pook/home/home.20160814T2050, receive=/mnt/data/backups/pook/home) At subvol /mnt/data/backup/pook/home/home.20160814T2050
ERROR: open /mnt/data/backups/pook/home/home.20160814T2050/o285733-1254589-0 failed. No such file or directory
WARNING: [send/receive] (send=/mnt/data/backup/pook/home/home.20160814T2050, receive=/mnt/data/backups/pook/home) At snapshot home.20160814T2050
ERROR: Failed to send/receive btrfs subvolume: /mnt/data/backup/pook/home/home.20160814T2050 [/mnt/data/backup/pook/home/home.20160807T2140] -> fukpi:/mnt/data/backups/pook/home
WARNING: Deleted partially received (garbled) subvolume: fukpi:/mnt/data/backups/pook/home/home.20160814T2050
ERROR: Error while cloning, aborting
WARNING: Skipping archiving of "/mnt/data/backup/pook/home/": At least one target aborted earlier
```

I just tried the command again, and it does not fail right away: it creates a new snaphot on the archive target and starts to fill it. Strangely, many directories appear inside that new snapshot, all with names like `o257-1306960-0`, but maybe that's part of normal operations.
UPDATE: after transferring data for many hours, the next try failed as well, with the same error message. So I'm really stuck here.

Any idea what could have gone wrong here? One thing to note is that this is a long running operation (these snapshots are pretty big (~50G) and need to be transferred over the internet), and while it is running, new snapshots are being created (and maybe some of them deleted) by the primary backup process – could that possibly interfere with the archive process?

**Comment by mac-linux-free** [comment]
Hi, I have the same issue sometimes. But now with btrfs-progs-4.7.3 and the latest kernel 4.7.5 they are gone till now.

**Comment by ariep** [comment]
Thanks for the info!
I'm hesitant to depart from debian stable on these machines, but if no other solutions come up I may try to upgrade.

**Comment by mac-linux-free** [comment]
a good alternative is the manjaro-net edition. I do not want to miss the actuality.

**Comment by ariep** [comment]
I kept the kernel at version 4.4.15 for now, but compiled a fresh btrfs-progs from git, and restarted the `btrbk archive` operation. (From a quick look at the source, it appears btrbk picks up the btrfs command from the $PATH, and is not hard-coded to `/sbin/btrfs`.)
Let's see if this will work! It will take some hours before I know.

**Comment by digint** [comment]
@ariep your quick look at the code was too quick, btrbk [still hardcodes your PATH](https://github.com/digint/btrbk/blob/master/btrbk#L3604) (this has caused some confusion and is subject to change in upcoming v0.24.0):
```
$ENV{PATH} = '/sbin:/bin:/usr/sbin:/usr/bin';
```

As I can tell from older posts, this kind of error is sometimes triggered by some bad source subvolume, and can only be resolved by deleting it (`btrfs sub delete /mnt/data/backup/pook/home/home.20160814T2050` in your case), as it will always produce a "buggy" send-stream. Note that the send-stream is generated by the kernel, and the btrfs-progs have not changed much in the receiving part lately. This means your problem _might_ get resolved when using latest kernel on the source side.

**Comment by ariep** [comment]
Thanks for the correction on the $PATH, that explains why upgrading btrfsprogs didn't have any effect ;-) (in addition to possibly not even helping even if the new version is used, as you mention).

I'll delete the problematic snapshot and try again. Even if that works this isn't a permanent solution, as such faulty snapshots could be created again by the same process. But if I understand you correctly, upgrading the kernel might prevent such faulty snapshots from being created?

**Comment by digint** [comment]
Having an up-to-date kernel is a good thing when using btrfs, as this filesystem is still pretty young and still "stabilising". So yes, upgrading the kernel _might_ prevent this to happen again.

I agree that deleting the troublesome snapshots is not really a desired solution, but sadly it's the only solution I know of. Maybe re-balancing helps, but this is also not satisfying. Maybe the btrfs devs can help you investigating this issue and offer you a better solution (ask on freenode or on the mailing list, running the latest kernel is always appreciated there).

**Comment by ariep** [comment]
Deleting the single snapshot didn't help: the next snapshot that was to be transferred also failed.
Interestingly, the error message is almost the same, including the name of the "missing" file `/mnt/data/backups/pook/home/home.20160821T1150/o285733-1254589-0`. Weird. What are these files, anyway (I have no files named like that)?

About the up-to-date kernel: sure, and three or four years back when I started to experiment with btrfs, I updated to the latest kernel version continuously. I reckoned that by now, at least this part of btrfs had stabilised, but apparently not.

**Comment by digint** [comment]
The btrfs send stream is basically a chain of unix commands ("create file X", "insert data to file Y at offset Z", ...) creating identical data structures at the receiving target. So at one point in the send stream, there is the command "open o285733-1254589-0", wrongly assumed that this file was created earlier in the stream. I don't know the details, but I assume that this file will probably be renamed to a file you really have in a later step. Or maybe btrfs somehow thinks you really have such a file, which may be some temporary file created at one point which was not correctly handled by btrfs.

Have you tried to run `btrfs check --readonly`? (use latest btrfs-progs version for this, there has been some improvements, and note that running it with `--repair` can cause more damage)

Ah and by the way, instead of deleting a buggy subvolume you could rename it (simply `mv` it to a different name), and btrbk will ignore it. Let's hope that at some point the subvolumes are healthy again...

**Comment by ariep** [comment]
It's been a long while; let me summarise what happened for those following along at home.

I tried to compile a newer kernel for my raspberry pi, but it wouldn't boot from it; I'm not sure why.

Some time later I deleted all relevant subvolumes from the primary backup computer. After btrbk had recreated (transferred) these subvolumes on the primary backup computer from the source computer, `btrbk archive` succeeded in transferring these subvolumes to the secondary backup. Apparently either my manual deletion of the primary backup subvolumes, or perhaps the weeding out over time of some subvolumes on the source according to the retention policy, removed the problematic snapshot(s).

Anyway, archiving my backups now works, so I'm a happy camper. I'll close this issue for now, but may reopen if this problem reoccurs.

Thanks for all your help!
